### PR TITLE
Split Row into CheckedRow and UncheckedRow

### PR DESCRIPTION
**Pull request by cmelchior**
Currently our typed API uses Row as the backing wrapper object for all Row access. This object does input validation in JNI for fieldnames/types on each access. This is wasted work since all those checks are done upfront when opening the Realm (validateTable).

This PR paves the way for the Dynamic API ( #1200 )  by splitting Row into UncheckedRow and CheckedRow. The Dynamic API will use CheckedRow, which will include those checks, while our typed API has moved to UncheckedRow.  

Row has been converted to an interface.

**Performance test:**
I did an informal test by running RealmResultsTest. The average running time decreased by 20%, which means the gain for purely the getters/setters have been even higher.

@emanuelez @kneth @nhachicha

### CODE REVIEW

File: realm-jni/src/io_realm_internal_CheckedRow.cpp [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_CheckedRow_nativeGetColumnIndex
+  (JNIEnv* env, jobject obj, jlong nativeRowPtr, jstring columnName)
+{
+    if (!ROW(nativeRowPtr)->is_attached())
+        return 0;
+
+    jlong ndx = Java_io_realm_internal_UncheckedRow_nativeGetColumnIndex(env, obj, nativeRowPtr, columnName);
+    if (ndx == realm::not_found) {
```
Comment (kneth) [inline realm-jni/src/io_realm_internal_CheckedRow.cpp]: I'm not sure if it's a problem but `realm::not_found` is `size_t(-1)` while `ndx` will have the value `jlong(-1)`. `jlong` is always a 64 bit signed integer while `size_t` is unsigned and depends on bit width.

File: realm/src/androidTest/java/io/realm/IOSRealmTests.java [hunk 2]
```diff
@@ -1,3 +1,18 @@
+/*
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/IOSRealmTests.java]: Unrelated fix

File: realm-jni/src/io_realm_internal_CheckedRow.cpp [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    if (!ROW_AND_COL_INDEX_AND_TYPE_VALID(env, ROW(nativeRowPtr), columnIndex, type_Double))
+        return;
+
+    Java_io_realm_internal_UncheckedRow_nativeSetDouble(env, obj, nativeRowPtr, columnIndex, value);
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_CheckedRow_nativeSetDate
+  (JNIEnv* env, jobject obj, jlong nativeRowPtr, jlong columnIndex, jlong value)
+{
+    if(!ROW_AND_COL_INDEX_AND_TYPE_VALID(env, ROW(nativeRowPtr), columnIndex, type_DateTime))
```
Comment (kneth) [inline realm-jni/src/io_realm_internal_CheckedRow.cpp]: `if (!`

File: realm/src/main/java/io/realm/internal/CheckedRow.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ * limitations under the License.
+ *
+ */
+
+package io.realm.internal;
+
+/**
+ * Checked wrapper for Row data in Realm Core. All methods called through this will check that input parameters are
+ * valid.
+ *
```
Comment (kneth) [inline realm/src/main/java/io/realm/internal/CheckedRow.java]: `<p/>` missing.

File: realm/src/main/java/io/realm/internal/StubRow.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+ * limitations under the License.
+ *
+ */
+
+package io.realm.internal;
+
+import java.util.Date;
+
+/**
+ * Row wrapper that stubs all access with IllegalStateExceptions. This can be used instead of adding null checks
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/internal/StubRow.java]: Nice idea!

File: realm/src/main/java/io/realm/internal/CheckedRow.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    private CheckedRow(UncheckedRow row) {
+        super(row.context, row.parent, row.nativePointer);
+        this.originalRow = row;
+    }
+
+    /**
+     * Get the row object associated to an index in a Table.
+     * @param context the Realm context.å
```
Comment (kneth) [inline realm/src/main/java/io/realm/internal/CheckedRow.java]: Remove "å"

File: realm/src/main/java/io/realm/internal/StubRow.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        throw getStubException();
+    }
+
+    @Override
+    public boolean isAttached() {
+        throw getStubException();
+    }
+
+    private RuntimeException getStubException() {
+        return new IllegalStateException("Object is no longer manged by Realm. Has it been deleted?");
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/internal/StubRow.java]: managed

File: realm/src/main/java/io/realm/internal/UncheckedRow.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2014 Realm Inc.
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/internal/UncheckedRow.java]: 2015

File: realm/src/main/java/io/realm/internal/LinkView.java [hunk 9]
```diff
@@ -33,8 +33,28 @@ public LinkView(Context context, Table parent, long columnIndexInParent, long na
         this.nativeLinkViewPtr = nativeLinkViewPtr;
     }
 
-    public Row get(long pos) {
-        return Row.get(context, this, pos);
+    /**
+     * Returns a non-checking Row. Incorrect use of this Row will cause a hard Realm Core crash.
+     * If error checking is required, use {@link #getCheckedRow(long)} instead.
```
Comment (kneth) [inline realm/src/main/java/io/realm/internal/LinkView.java]: I think we should emphasise that we recommend the checked Row.

File: realm/src/main/java/io/realm/internal/LinkView.java [hunk 10]
```diff
@@ -33,8 +33,28 @@ public LinkView(Context context, Table parent, long columnIndexInParent, long na
...
+     *
+     * @param index Index of row to fetch.
+     * @return Unsafe row wrapper object.
+     */
+    public UncheckedRow getUncheckedRow(long index) {
+        return UncheckedRow.get(context, this, index);
+    }
+
+    /**
+     * Returns a wrapper around Row access. All access will be error checked in JNI and will throw an
```
Comment (kneth) [inline realm/src/main/java/io/realm/internal/LinkView.java]: "around" -> "for"

File: realm/src/main/java/io/realm/internal/UncheckedRow.java (hunk 8 above)
Comment (kneth) [inline realm/src/main/java/io/realm/internal/UncheckedRow.java]: "2014" -> "2015"

File: realm/src/main/java/io/realm/internal/UncheckedRow.java [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+ */
+
+package io.realm.internal;
+
+import java.util.Date;
+
+/**
+ * Wrapper around a Row in Realm Core.
+ *
+ * IMPORTANT: All access to methods using this class are non-checking. Safety guarantees are given by the the annotation
```
Comment (kneth) [inline realm/src/main/java/io/realm/internal/UncheckedRow.java]: "the the" -> "the"

File: realm/src/androidTest/java/io/realm/IOSRealmTests.java (hunk 2 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/IOSRealmTests.java]: Sorry

**Comment by cmelchior** [comment]
All issues resolved: @emanuelez @kneth

**Comment by kneth** [comment]
:+1:

File: realm/src/main/java/io/realm/internal/InvalidRow.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+package io.realm.internal;
+
+import java.util.Date;
+
+/**
+ * Row wrapper that stubs all access with IllegalStateExceptions. This can be used instead of adding null checks
+ * everywhere when the underlying Row accessor in Realm Core is no longer available.
+ */
+public class InvalidRow implements Row {
```
Comment (nhachicha) [inline realm/src/main/java/io/realm/internal/InvalidRow.java]: Should be final (to avoid accidentally extending) & Probably singleton since this is immutable (no need to create a new instance for each InvalidRow).

File: realm/src/main/java/io/realm/internal/CheckedRow.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ */
+
+package io.realm.internal;
+
+/**
+ * Checked wrapper for Row data in Realm Core. All methods called through this will check that input parameters are
+ * valid or throw an appropriate exception otherwise.
+ * <p>
+ * For low-level access to a Realm where safety is not a concern use {@link UncheckedRow} instead for improved
```
Comment (nhachicha) [inline realm/src/main/java/io/realm/internal/CheckedRow.java]: _safety is not a concern_  because checks were already performed :)

**Comment by nhachicha** [comment]
:+1:

**Comment by emanuelez** [comment]
:+1:
